proposals/0000-replacing-cabal-custom-build.md

```diff
@@ -0,0 +1,1378 @@
+
+
+#### Phase separation
+
+Only configure hooks can make changes to the `PackageDescription`.  Once
+configuration is finished, the package description should be set in stone, and
+subsequent hooks such as build hooks are not able to modify it.
+This differs from `UserHooks`, where modifications to the project configuration
+in the configure phase are not propagated to the other phases, and instead
+subsequent hooks must re-apply changes, in the form of `HookedBuildInfo`.
+This old design lead to several subtle bugs and maintenance headaches which
+this new design will allow us to get rid of, once we remove support for
+the `Custom` build-type.
+
+The configuration hooks thus follow a simple philosophy:
+
+* All modifications to global package options must use `preConfPackageHook`.
+* All modifications to component configuration options must use `preConfComponentHook`.
+
+If a hook modifies the options in these phases, then the configuration is propagated
+into all subsequent phases, and the design of the interface ensures that
+this is the only point where hooks can modify the options.
+
+It is only the pre-configuration hooks which allow modification of the options.
+This is because the configuration process computes some more complicated data
+structures from these initial inputs. If hooks were allowed to modify the results
+of configuration then it would be error-prone to ensure that they suitably updated
+both the options in question as well as the generated configuration.
+For example, both `PackageDescription` and `ComponentLocalBuildInfo` contain
+a list of exposed modules for the library.
+This is why the "post" configuration hooks (and any hooks after that)
+can only run an `IO` action; they can't return any modifications that would affect the `PackageDescription`.
+
+
+#### `LocalBuildConfig`
+
+There are parts of the `LocalBuildInfo` which must be decided at a global (per-package) level.
+For instance, whether to build dynamic libraries.
+On the other hand, there are also things we want to decide on a local (per-component) level,
+such as specific GHC options with which to compile the component.
+
+Moreover, there are parts of the `LocalBuildInfo` which hooks cannot modify.
+For example, things such as package dependencies can't be modified because they are
+determined externally by the overall build plan (e.g. from the dependency solver).
+Thus the hooks interface should prevent the modification of these
+parts of `LocalBuildInfo`.
+
+We propose to achieve this by defining a new type `LocalBuildConfig` which
+contains only the parts of the existing `LocalBuildInfo` datatype that can be
+modified by `preConfPackageHook`.
+
+
+#### `ComponentDiff`
+
+The `ComponentDiff` records the modifications that should be applied to each component.
+
+For each component, `preConfComponentHook` is run, returning a `ComponentDiff`.
+This `ComponentDiff` is applied to its corresponding `Component`
+by monoidally combining together the fields.
```

Comment:
I actually sort of like the monoidal combination -- it means that preConf hooks can only _add_ data to components, not delete it. And also the content of the diff itself is then easily statically inspectable.

Reply:
I guess part of my objection is that it's inconsistent. E.g. the pre-conf hook returns a new `LocalBuildInfo`, not a separate diff type!

Reply:
The type of the package-wide pre-configure hook is, more or less, `LocalBuildConfig -> IO LocalBuildConfig`, where `LocalBuildConfig` extracts out parts of `LocalBuildInfo` that are package-wide and that it makes sense for the user to modify. The reason it isn't a diff type is because the monoidal structure is a bit more complicated: you might want to override a field, or you might want modify its value in different ways. To model this with a diff type would be a bit more complicated. For example, we could wrap each field in `data ModifyOrOverride a = Don'tTouch | Override a | Combine a` to define the diff type, but that seems a bit heavyweight.

This is different for components: at least with their current definition, the only thing one does in practice is to append additional data to certain fields. And returning a diff, rather than an arbitrary function, makes it easier to statically analyse what change the hook applies.

Reply:
That makes sense as an explanation but leaves me uneasy. What happens if we do change the definition of `Component`? Not having the more powerful form of modification seems less future-proof.

Reply:
@gbaz Some examples are overriding the `OptimisationLevel` field, overriding programs in the `ProgramDb`, and overriding command-line arguments rather than appending to them. I'll admit it's not necessarily very compelling, as there are a lot of fields one in practice never touches, and it might be a good argument to slim down the interface as John has been advocating (in parallel with the whole `DynFlags` saga).

Reply:
I'm still somewhat unsatisfied with the argument here. As a user of the API it would strike me as weird that these two conceptually very similar things were so different.